inspect_dynspec/gpr_smooth.py

```diff
@@ -0,0 +1,266 @@
+                is the hyperplane approximation of y and X is x
+                with a row of ones appended as the final axis
+    """
+    D, N = x.shape
+    y = y.squeeze()[None, :]
+    z = jnp.vstack((x, y))
+    centroid = jnp.zeros((D + 1, 1), dtype=y.dtype)
+    for d in range(D + 1):
+        if d < D:
+            centroid = centroid.at[d, 0].set(jnp.sum(x[d]) / N)
+        else:
+            centroid = centroid.at[d, 0].set(jnp.sum(y) / N)
+    diff = z - centroid
+    cov = diff.dot(diff.conj().T)
+    s, V = jnp.linalg.eigh(cov)
+    n = V[:, 0].conj()  # defines normal to the plane
+    theta = jnp.zeros(D + 1, dtype=y.dtype)
+    for d in range(D + 1):
+        if d < D:
+            theta = theta.at[d].set(-n[d] / n[-1])
+        else:
+            # we need to take the mean here because y can be noisy
+            # i.e. we do not have a point exactly in the plane
+            theta = theta.at[d].set(jnp.mean(n[None, 0:-1].dot(x) / n[-1] + y))
+    return theta
+
+
+def kron_mv(Ls : tuple, z : jnp.ndarray) -> jnp.ndarray:
+    """
+    Generalized Kronecker matvec for JAX, matching the utils.py kron_matvec logic.
+    Ls: list of matrices (e.g. [Lv, Lt])
+    z: flattened input vector
+    """
+    x = z
+    for A in Ls:
+        Gd = A.shape[0]
+        NGd = x.size // Gd
+        X = x.reshape(Gd, NGd)
+        Z = A @ X
+        x = Z.T.ravel()
+    return x.reshape(z.shape)
+
+
+def rbf_kernel(grid : jnp.ndarray, lengthscale: float, variance: float=1.0) -> jnp.ndarray:
+    """
+    RBF kernel K_ij = variance * exp(-0.5 * (xi - xj)^2 / lengthscale^2)
+    """
+    d2 = jnp.subtract.outer(grid, grid) ** 2
+    return variance * jnp.exp(-0.5 * d2 / lengthscale**2)
+
+
+class Mask:
+    def __init__(self, mask: jnp.ndarray):
+        """
+        A JAX-compatible mask operator.
+        mask: boolean array of shape (nx, ny), True for observed pixels.
+        """
+        self.shape = mask.shape
+        # Flattened boolean mask
+        self.mask_flat = mask.ravel()
+        # Precompute number of observations
+        self.n_obs = int(self.mask_flat.sum())
+
+    def forward(self, x: jnp.ndarray) -> jnp.ndarray:
+        """
+        R @ x: pick out observed pixels based on the mask
+        x can be shape (nx, ny) or flattened (nx*ny,) - costs nothing to flatten if already flattened
+        """
+        x_flat = x.ravel()
+        return x_flat[self.mask_flat]  # shape (n_obs,)
+
+    def adjoint(self, y_obs: jnp.ndarray) -> jnp.ndarray:
+        """
+        R_T @ y_obs: scatter residuals back into full image.
+        Returns array of shape (nx * ny,), which you can reshape to (nx, ny).
+        """
+        # start with zeros in flattened space
+        full = jnp.zeros(self.mask_flat.shape)
+        # scatter observed values back
+        full = full.at[self.mask_flat].set(y_obs)
+        return full  # still flattened
+
+    def adjoint_image(self, y_obs: jnp.ndarray) -> jnp.ndarray:
```

Comment:
Is there a good reason for adding the reshape as a separate method? If not, consider making it part of `adjoint` so that it really is the adjoint of the `forward` operator